File: python/scenario/voice/tts.py
"""
Text-to-speech router and cache.

The TTS side uses litellm-style ``provider/voice_name`` routing (e.g.
``openai/nova``, ``elevenlabs/rachel``). Per the TTS cache key locked decision
the cache key is ``(text, voice)`` only; audio effects are applied AFTER a
cache hit and are never baked into the cached audio.

Users can register additional providers via ``register_tts_provider(...)``.
The default set covers OpenAI (hard dep) and lazy-imports ElevenLabs /
Google / Cartesia only when their provider prefix is actually used.
"""

from __future__ import annotations

import asyncio
import functools
from dataclasses import dataclass
from typing import Awaitable, Callable, Dict, Tuple

from ..cache import scenario_cache
from .audio_chunk import AudioChunk, PCM16_SAMPLE_RATE


TTSCallable = Callable[[str, str], Awaitable[bytes]]
"""(text, voice_name) -> PCM16 @ 24kHz mono bytes"""


_PROVIDERS: Dict[str, TTSCallable] = {}


def register_tts_provider(prefix: str, synth: TTSCallable) -> None:
    """Register a TTS backend under the given provider prefix."""
    _PROVIDERS[prefix.lower()] = synth


def _split_voice(voice: str) -> Tuple[str, str]:
    if "/" not in voice:
        raise ValueError(
            f"Voice string {voice!r} must be in 'provider/name' format, e.g. 'openai/nova'"
        )
    provider, name = voice.split("/", 1)
    return provider.lower(), name


# ---------------------------------------------------------------- default TTS

async def _openai_tts(text: str, voice: str) -> bytes:
    """Default OpenAI TTS provider. Uses gpt-4o-mini-tts for short clips."""
    import io
    from openai import AsyncOpenAI

    client = AsyncOpenAI()
    # PCM format means raw PCM16 @ 24kHz mono — matches our internal AudioChunk.
    response = await client.audio.speech.create(
        model="gpt-4o-mini-tts",
        voice=voice,
        input=text,
        response_format="pcm",
    )
    audio_bytes = await response.aread() if hasattr(response, "aread") else response.read()
    return audio_bytes


def _register_default_providers() -> None:
    register_tts_provider("openai", _openai_tts)
    # Lazy imports — only attempt the import when the prefix is used.
    async def _elevenlabs(text: str, voice: str) -> bytes:  # noqa: E306
        try:
            from elevenlabs.client import AsyncElevenLabs  # type: ignore
        except ImportError as exc:
            raise ImportError(
                "elevenlabs provider requires `pip install elevenlabs`"
            ) from exc
        client = AsyncElevenLabs()
        chunks = []
        async for chunk in await client.text_to_speech.convert(
            voice_id=voice,
            text=text,
            output_format="pcm_24000",
        ):
            chunks.append(chunk)
        return b"".join(chunks)

    register_tts_provider("elevenlabs", _elevenlabs)

    async def _google(text: str, voice: str) -> bytes:  # noqa: E306
        try:
            from google.cloud import texttospeech  # type: ignore
        except ImportError as exc:
            raise ImportError(
                "google provider requires `pip install google-cloud-texttospeech`"
            ) from exc
        client = texttospeech.TextToSpeechAsyncClient()
        synth_input = texttospeech.SynthesisInput(text=text)
        voice_cfg = texttospeech.VoiceSelectionParams(
            language_code="-".join(voice.split("-")[:2]) or "en-US",
            name=voice,
        )
        audio_cfg = texttospeech.AudioConfig(
            audio_encoding=texttospeech.AudioEncoding.LINEAR16,
            sample_rate_hertz=PCM16_SAMPLE_RATE,
        )
        resp = await client.synthesize_speech(
            input=synth_input, voice=voice_cfg, audio_config=audio_cfg
        )
        return bytes(resp.audio_content)

    register_tts_provider("google", _google)

    async def _cartesia(text: str, voice: str) -> bytes:  # noqa: E306
        try:
            from cartesia import AsyncCartesia  # type: ignore
        except ImportError as exc:
            raise ImportError(
                "cartesia provider requires `pip install cartesia`"
            ) from exc
        client = AsyncCartesia()
        raw = await client.tts.bytes(
            model_id="sonic-english",
            transcript=text,
            voice_id=voice,
            output_format={
                "container": "raw",
                "encoding": "pcm_s16le",
                "sample_rate": PCM16_SAMPLE_RATE,
            },
        )
        return raw


_register_default_providers()


# ------------------------------------------------------------ cached synthesis

@dataclass(frozen=True)
class _TTSKey:
    text: str
    voice: str


async def _synthesize_raw(text: str, voice: str) -> bytes:
    provider, name = _split_voice(voice)
    if provider not in _PROVIDERS:
        raise ValueError(
            f"Unknown TTS provider {provider!r}. Known: {sorted(_PROVIDERS)}"
        )
    return await _PROVIDERS[provider](text, name)


# joblib cache is sync, but we only call synth once per (text, voice). Cache
# at the PCM-bytes level keyed exclusively on (text, voice) — per the locked
# decision, effects are NEVER baked into the cached audio.
@scenario_cache()
def _cached_synthesize_sync(text: str, voice: str) -> bytes:
    loop = asyncio.new_event_loop()
    try:
        return loop.run_until_complete(_synthesize_raw(text, voice))
    finally:
        loop.close()


async def synthesize(text: str, voice: str) -> AudioChunk:
    """
    Synthesize ``text`` into an AudioChunk using the voice provider.

    Cache key is ``(text, voice)``. Effects must be applied by the caller on
    the returned chunk — they are never part of the cache key.
    """
    # Run the sync cached call in a thread so we don't block the event loop.
    pcm = await asyncio.to_thread(_cached_synthesize_sync, text, voice)
    return AudioChunk(data=pcm)

File: javascript/src/voice/tts/tts.ts
/**
 * Text-to-speech router and cache (the core; provider leaves live alongside).
 *
 * Python parity: `python/scenario/voice/tts.py`. Litellm-style routing — voice
 * strings are `provider/name` (e.g. `openai/nova`, `elevenlabs/rachel`). The
 * TTS cache key is `(sha256(text), voice, voiceStyle)` so raw user-supplied
 * text never reaches the cache payload; audio effects apply AFTER cache hit
 * and are never baked into stored audio.
 *
 * TS leads Python here: `voiceStyle` (issue #533) is a per-call synthesis
 * option threaded from the user simulator down to the provider. Python's
 * `synthesize(text, voice)` still has no style channel — this is NOT parity.
 *
 * Concrete providers (OpenAI, ElevenLabs) are one-file-per-provider leaves that
 * self-register via `tts/index.ts` (mirrors the `stt/` subtree, EDR §5.3). This
 * module owns the interface, the registry router, `synthesize()`, and the LRU
 * cache only — no provider SDK imports.
 */
import { createHash } from "node:crypto";

import { AudioChunk } from "../audio-chunk";

/**
 * Per-call synthesis options that are NOT part of the voice identity.
 *
 * Kept separate from the `provider/name` voice string because the SAME voice
 * can speak in many styles — the style varies per utterance, the voice does
 * not. Optional on {@link TTSCallable} so existing two-argument provider
 * callables stay assignable.
 */
export interface TtsSynthesisOptions {
  /**
   * Named delivery style for this utterance, e.g. `"angry"`, `"whispering"`.
   * Providers map it onto their own style channel (ElevenLabs: an inline
   * `[angry]` marker; OpenAI: the `instructions` parameter) and opt in via
   * {@link TtsProvider.supportsVoiceStyle}. A provider that has not opted in
   * gets the style stripped, and the router warns once.
   */
  voiceStyle?: string;
}

/**
 * A TTS backend: takes (text, voiceName, options?) and returns PCM16/24kHz
 * mono bytes. `options` carries per-call, non-identity knobs such as
 * {@link TtsSynthesisOptions.voiceStyle}.
 */
export type TTSCallable = (
  text: string,
  voiceName: string,
  options?: TtsSynthesisOptions,
) => Promise<Uint8Array>;

/**
 * A TTS provider registration — a litellm-style prefix and the function that
 * synthesizes for the names served by that prefix.
 */
export interface TtsProvider {
  prefix: string;
  synth: TTSCallable;
  /**
   * Whether `synth` honours {@link TtsSynthesisOptions.voiceStyle}. Defaults
   * to `false` — a provider that has not opted in has the style STRIPPED and
   * the router warns once, rather than quietly returning unstyled audio the
   * caller believes is styled.
   */
  supportsVoiceStyle?: boolean;
}

/**
 * Apply post-cache audio shaping. Effects are pure functions over the canonical
 * AudioChunk; they never participate in the cache key.
 */
export type TtsEffectFn = (chunk: AudioChunk) => AudioChunk | Promise<AudioChunk>;

/** Registry entry — the callable plus the capabilities it declared. */
interface RegisteredTtsProvider {
  synth: TTSCallable;
  supportsVoiceStyle: boolean;
}

const PROVIDERS = new Map<string, RegisteredTtsProvider>();

/**
 * In-process LRU cache keyed on (sha256(text), voice, voiceStyle) → PCM16
 * bytes. Bounded to prevent unbounded memory growth — a 5-minute clip is
 * ~14 MB, so 64 entries caps the cache at ~900 MB even for long utterances.
 * (Mirrors the Python tuning.)
 */
const CACHE_MAX_ENTRIES = 64;
const CACHE = new Map<string, Uint8Array>();

/** Clear the in-process TTS cache. Used by tests and long-lived processes. */
export function clearTtsCache(): void {
  CACHE.clear();
}

/**
 * Prefixes already warned about an ignored `voiceStyle`. One warning per
 * provider for the life of the process — a styled multi-turn run would
 * otherwise repeat the same line on every turn.
 */
const VOICE_STYLE_WARNED = new Set<string>();

/**
 * Reset the one-shot "provider ignores voiceStyle" warning ledger.
 *
 * @internal Test-only seam. Deliberately NOT folded into
 * {@link clearTtsCache}: the two have unrelated lifetimes, and a suite that
 * clears the cache between cases must not silently re-arm the warning it is
 * asserting fires exactly once.
 */
export function __resetVoiceStyleWarnings(): void {
  VOICE_STYLE_WARNED.clear();
}

function warnVoiceStyleUnsupportedOnce(
  provider: string,
  voiceStyle: string,
): void {
  if (VOICE_STYLE_WARNED.has(provider)) return;
  VOICE_STYLE_WARNED.add(provider);
  console.warn(
    `[scenario] TTS provider ${JSON.stringify(provider)} does not support ` +
      `voiceStyle — the style ${JSON.stringify(voiceStyle)} is ignored and ` +
      "this turn is synthesized unstyled. Providers declare support with " +
      "registerTtsProvider({ …, supportsVoiceStyle: true }).",
  );
}

/** Register a TTS backend under the given provider prefix. */
export function registerTtsProvider(provider: TtsProvider): void {
  PROVIDERS.set(provider.prefix.toLowerCase(), {
    synth: provider.synth,
    supportsVoiceStyle: provider.supportsVoiceStyle ?? false,
  });
}

/** Test-only: enumerate registered provider prefixes. */
export function listTtsProviders(): string[] {
  return Array.from(PROVIDERS.keys()).sort();
}

function splitVoice(voice: string): { provider: string; name: string } {
  const slash = voice.indexOf("/");
  if (slash < 0) {
    throw new Error(
      `Voice string ${JSON.stringify(voice)} must be in 'provider/name' format, ` +
        "e.g. 'openai/nova'",
    );
  }
  return {
    provider: voice.slice(0, slash).toLowerCase(),
    name: voice.slice(slash + 1),
  };
}

function hashText(text: string): string {
  return createHash("sha256").update(text, "utf8").digest("hex");
}

function cacheKey(textHash: string, voice: string, voiceStyle?: string): string {
  // Composite key — text hash, voice and voice style are all load-bearing.
  // "voice" is the full provider/name string so two providers can't collide.
  // `voiceStyle` is in the key because it changes the synthesized AUDIO, not
  // just the request: without it, an angry turn and a neutral turn with the
  // same (text, voice) would share one entry, so whichever ran first would
  // silently serve its bytes to the other (issue #533). Effects stay OUT of
  // the key — they are applied after the cache read, never baked in.
  return `${textHash}:${voice}:${voiceStyle ?? ""}`;
}

/** True when `options` carries at least one set value (all fields optional). */
function hasAnyOption(options?: TtsSynthesisOptions): boolean {
  return (
    options !== undefined && Object.values(options).some((v) => v !== undefined)
  );
}

async function synthesizeRaw(
  text: string,
  voice: string,
  options?: TtsSynthesisOptions,
): Promise<Uint8Array> {
  const { provider, name } = splitVoice(voice);
  const registered = PROVIDERS.get(provider);
  if (!registered) {
    throw new Error(
      `Unknown TTS provider ${JSON.stringify(provider)}. Known: ${listTtsProviders().join(", ") || "(none)"}`,
    );
  }

  // A style the backend cannot honour must be loud, not silent — otherwise a
  // scripted `user("…", { voiceStyle: "angry" })` produces cheerful audio and
  // nothing says why. Warn once per provider, then strip ONLY the style so
  // any other per-call option still reaches the backend.
  let effective = options;
  const voiceStyle = options?.voiceStyle;
  if (voiceStyle && !registered.supportsVoiceStyle) {
    warnVoiceStyleUnsupportedOnce(provider, voiceStyle);
    effective = { ...options, voiceStyle: undefined };
  }

  // Providers registered before per-call options existed are two-argument
  // callables. Only widen the call when an option is actually set, so the
  // (dominant) unstyled path keeps its historical arity — a third `undefined`
  // argument is observable to callables that inspect `arguments.length`.
  return hasAnyOption(effective)
    ? registered.synth(text, name, effective)
    : registered.synth(text, name);
}

/**
 * Synthesize `text` into an {@link AudioChunk} using the voice provider.
 *
 * Cache key is `(sha256(text), voice, voiceStyle)` — equivalent determinism to
 * `(text, voice, voiceStyle)` without pinning raw text in the cache payload.
 * Effects pass through `effectFn` AFTER a cache hit and are never part of the
 * key, matching the locked-decision invariant from the Python port.
 *
 * @param options Per-call synthesis knobs (currently
 *   {@link TtsSynthesisOptions.voiceStyle}) forwarded to the provider.
 */
export async function synthesize(
  text: string,
  voice: string,
  effectFn?: TtsEffectFn,
  options?: TtsSynthesisOptions,
): Promise<AudioChunk> {
  const key = cacheKey(hashText(text), voice, options?.voiceStyle);
  let pcm = CACHE.get(key);
  if (pcm !== undefined) {
    // LRU touch — delete + set re-inserts at the tail of insertion order.
    CACHE.delete(key);
    CACHE.set(key, pcm);
  } else {
    pcm = await synthesizeRaw(text, voice, options);
    CACHE.set(key, pcm);
    while (CACHE.size > CACHE_MAX_ENTRIES) {
      const oldest = CACHE.keys().next().value;
      if (oldest === undefined) break;
      CACHE.delete(oldest);
    }
  }
  const chunk = new AudioChunk({ data: pcm, transcript: text });
  if (effectFn) {
    return effectFn(chunk);
  }
  return chunk;
}
